### ISSUE TITLE
v2.2

### ISSUE THREAD

**Issue by Arlen22** [someguynamedmatt/TiddlyServer#20]
Hi Matt, 

just wanted to make sure you know about the v2.2 branch. It's now the default branch (aka master) and I'm probably going to release the first version shortly. I've already pushed it to npm, and I've gotten the whole thing working with webpack. Feel free to check it out. The build.sh file does everything needed to build the webpack version and publish to npm, so if you run that you will get the npm equivalent. 

You can run `npm install -g tiddlyserver` if you want to try it out, but more likely you'll just want to clone the v2.2 branch and run `./build.sh` from the root directory. I'd be happy to hear your feedback on it. It should all just work. 

Also, as far as converting the login page to react, the reason I decided against that was because it had exactly one function and it seemed like react was overkill for such a simple UI, but hey, feel free. 

Also, I am worried that it might be confusing for end users to have a fork with only changes under the hood, so I'm hoping that we can merge our work rather than creating a TiddlyServer fork that is almost identical in functionality from the end user perspective. I do know the feeling of wanting to rearrange code and I am quite pleased that someone has taken the time to figure out my densely formatted code.

**Comment by someguynamedmatt** [comment]
@Arlen22 Yeah, definitely! That's my hope too. It'd be better, in general, if there was one canonical TiddlyServer. I'm not sure how much of what I'm doing you're following, but I'm trying to:

- consolidate everything into the `src` directory
- split everything into an intuitive `client`/`server` sub-directory structure (of `src`)
- remove dead code
- docker
  > I know we've briefly discussed the Docker situation before, but I think it would enable really quick deployments of this if someone wanted to put it up into the cloud

As for React, I'm not necessarily sold on doing that. It was an idea that, honestly, probably doesn't make a whole lot of sense. The one thing I would certainly like to do is clean up the client to the absolute bare minimum (I think there's some Angular code hanging around still) and have it be able to do its own redirects into/out-of the auth page.

**Comment by someguynamedmatt** [comment]
@Arlen22 I've only looked at the `build.sh` file so far, but I'm curious why that's been chosen over, say, putting those commands within the `package.json`?

**Comment by Arlen22** [comment]
> I've only looked at the build.sh file so far, but I'm curious why that's been chosen over, say, putting those commands within the package.json?

It felt like too many commands to keep organized in package.json. The lines were getting really long.

> - consolidate everything into the src directory
> - split everything into an intuitive client/server sub-directory structure (of src)

The server directory is currently `/src` and all the server code is consolidated in there now, except for the `server.ts` in the root, which is meant to be an executable file and so I'm planning to leave it there for now.

The client directory is currently `/assets/authenticate` and typescript compiles it to `/assets/static`. 

> remove dead code

There's not much dead code in`/src`. Well, I guess there could be, but I think the Webpack Terser probably removes it. I usually remove dead code once I'm sure I don't need it anymore. I'm cleaning up the `/assets` directory right now. 

> docker

Having an NPM install is probably the biggest thing required for a docker build. Now that that's in place, a docker container should be fairly simple.

**Comment by Arlen22** [comment]
I would prefer to keep the client-side code in assets, since there is very little of it and it almost never changes. Putting it in src feels like a lot of work. I'm open to your thoughts on that.

**Comment by Arlen22** [comment]
>  have it be able to do its own redirects into/out-of the auth page.

That should be some pretty simple code. If you have an idea for how it should be done feel free to modify `/assets/authenticate/login.ts`. I'm adding a redirect parameter to the login link in the directory index right now. I will also add it to the auth error login link. The login is done client-side so I don't think we should use a 302 Redirect on the server.

**Comment by Arlen22** [comment]
> I've only looked at the build.sh file so far, but I'm curious why that's been chosen over, say, putting those commands within the package.json?

I tried putting the commands in package.json but they felt kind of disorganized. I did set the start command to `tsc && node build/server --config settings.json` since that's the command I normally run to test my changes.

**Comment by Arlen22** [comment]
I've just done some of the things you mentioned but I'm not saying that's all I'm willing to do. It was just some easy stuff that I thought I could knock out while it was on my mind.

**Comment by someguynamedmatt** [comment]
@Arlen22 A few thoughts...

> I tried putting the commands in package.json but they felt kind of disorganized

Putting build/start scripts within the `package.json` is "best practice". It's not required, but considering this is an open-source, node package I think it's a good idea (for new developers that may want to help out with the project) to keep it in line with what most other packages, and the npm/javascript/node community do.

> I would prefer to keep the client-side code in assets, since there is very little of it and it almost never changes. Putting it in src feels like a lot of work. I'm open to your thoughts on that.

Please check out the master branch of this fork. I've moved the `assets` into `src/client`. Echoing the same sentiment above, it's not typical to have production code outside of the `src` directory. At least in my experience. I could be wrong, and this is certainly my opinion, but I don't think it lends much to the legibility of the code-base to have code scattered throughout various `src`-sibling directories.

___

I think I've done a lot of cleanup work that might be helpful to the `TiddlyServer` project as a whole. My intent with the fork was mostly that, to bring it inline with that standard node architectural patterns I've seen around. Please, at your convenience, take a look at my `master` branch. I've done a few things:

- removed a lot of code that wasn't being called anywhere else, but wasn't being flagged as "unused" because it was being exported

- moved `assets` into the `src` directory as a new directory (`client`)

- started shifting the server files (basically, what was all of the original `src/**/*`), but that's an ongoing thing and probably shouldn't be on my `master` branch

- added `nodemon` to the project so the development process is less taxing. In other words, someone can now run `npm run dev` and the server will be started and files will be watched and re-compiled into the running server (hot reloading) for a shorter feedback loop during development

- added a new webpack config (`webpack.client.ts`) this basically compiles/copies the `src` code into the `dist` directory. The original webpack config isn't used, and should be removed, but I haven't done that yet


I believe everything is still working as it should. I've really only tested it as far as authentication goes and that seems to be at parity with--at least--`v2.1`. Have a look and if you have any questions about it, feel free to ask. I really like `TiddlyServer` and would love to contribute my experience however I can.

**Comment by someguynamedmatt** [comment]
Oh, btw, if you want to get this thing up and running and see if it works correctly:

1. clone it

1. `npm i`

1. `npm run build && npm run start` (the caveat to this being the settings file, but I'm sure you're aware of that)

**Comment by Arlen22** [comment]
In looking over master this morning, I'm confused by webpack.client.ts. First of all, how can it be used if it is a TS file. I guess it get's compiled to `dist` along with everything else, and then the output is generated from there. Rather than having a separate build webpack for the client, I think it would be better to use the copy plugin in the server webpack config. 

If you're wanting to reorganize `/src` that's fine, but if you just want to move the files into `/src/server`, you can do that and as long as `tsc` doesn't give any errors, you're good to go. Do be aware that I'm also working on my own refactoring of the TiddlyServer code, so we'll probably be editing the same files. I'll look over what you've done so far in `/src/server` and try to understand the paradigm. 

I like what I'm seeing so far, as I do like things being standardized. I don't really know what best practices are for TypeScript projects, as I usually work by myself, so I'm glad for the advice you're giving.

**Comment by Arlen22** [comment]
> Putting build/start scripts within the package.json is "best practice". It's not required, but considering this is an open-source, node package I think it's a good idea (for new developers that may want to help out with the project) to keep it in line with what most other packages, and the npm/javascript/node community do.

I'll take another shot at it.

**Comment by Arlen22** [comment]
Ok, so what about putting the assets and cleanup in the webpack build? Would that work?

https://github.com/Arlen22/TiddlyServer/blob/56dd89fc10f1f316706c8e75cdfbded2cba25227/webpack.config.js

**Comment by someguynamedmatt** [comment]
> In looking over master this morning, I'm confused by webpack.client.ts. First of all, how can it be used if it is a TS file. I guess it get's compiled to dist along with everything else, and then the output is generated from there.

It's not moved anywhere. Webpack "knows" how to deal with configuration files [in other languages](https://webpack.js.org/configuration/configuration-languages/#typescript).

> Rather than having a separate build webpack for the client, I think it would be better to use the copy plugin in the server webpack config.

That file, `webpack.client.ts`, isn't meant to be long-lived. It's going to be a consolidated webpack configuration, I just left the original around for reference until I have it building/bundling correctly.

**Comment by someguynamedmatt** [comment]
> If you're wanting to reorganize /src that's fine, but if you just want to move the files into /src/server, you can do that and as long as tsc doesn't give any errors, you're good to go.

As of right now, there are no compilation errors. I do think, and this is a **huge** task, that creating a more type-safe environment would be ideal. What I mean by that is there's a lot of untyped/implicit typing going on. Specifically, I'm seeing a lot of `any`s everywhere. This isn't inherently bad, but considering this is supposed to be a server, I think it would be great if we had explicit types. Again, this is a much larger task that isn't important right now, but I just wanted to note it.

>  I'll look over what you've done so far in /src/server and try to understand the paradigm.

Right now there isn't too much of a paradigm to follow. What you're seeing on `master` is a migration from the old architecture to a more clearly delineated `src` directory that will breakdown the code base between the client-side code (what was `assets`) and the server-side code (what, originally, was all of `src`).

**Comment by someguynamedmatt** [comment]
> Ok, so what about putting the assets and cleanup in the webpack build? Would that work?

There's no need to reference the assets in the webpack config if we go with what I've done on my `master`. It's being copied over from `src/client` and is working. Cleanup can certainly be done in webpack.

**Comment by Arlen22** [comment]
https://github.com/Arlen22/TiddlyServer/blob/56dd89fc10f1f316706c8e75cdfbded2cba25227/webpack.config.js#L7-L8

How did I not know that ?!?!?!

Ok, I've updated v2.2, so feel free to check it out. I didn't move the server and client code yet as I'm waiting to see how you shake that all out.

> There's no need to reference the assets in the webpack config if we go with what I've done on my `master`. It's being copied over from `src/client` and is working. Cleanup can certainly be done in webpack.

Ok, check out my latest webpack.config.ts. I've consolidated everything into one webpack. We can still copy the client from the src directory instead of assets, but I'm pretty sure tsc won't copy it, so it has to be done in webpack. 

https://github.com/Arlen22/TiddlyServer/blob/v2.2/webpack.config.ts

Does "assets" mean something different than how I'm using it? I can see that the word "client" would fit, but I never think of it as having a client side, since it's not a client-side webapp or anything like that. But if it makes more sense to call it client, and put it in src, I'm fine with that.

**Comment by someguynamedmatt** [comment]
> I'm pretty sure tsc won't copy it

It will. I have it working on `master` I think. Maybe not, but I have it working somewhere :)

EDIT: Let me be more clear. The TypeScript is correctly transpiled into the `dist` directory, but I'm copying over the non-TS files into `dist`.

> Does "assets" mean something different than how I'm using it? I can see that the word "client" would fit, but I never think of it as having a client side, since it's not a client-side webapp or anything like that. 

Typically, "assets" means something like images or some non-code component to a project. At least that's been consistently true in my experience. While this isn't a "client side" app, that code is served specifically to a client, which is the meaning we want to convey to anyone else that may browse the source code. They'd see `src/client`, assume it was browser-side code, then come across its contents: HTML, CSS, and the JavaScript and the assumption would be validated. That's how I think of it anyway.

**Comment by Arlen22** [comment]
> > I'm pretty sure tsc won't copy it
> 
> It will. I have it working on `master` I think. Maybe not, but I have it working somewhere :)

I'll dig around. That'd be news to me (which wouldn't surprise me in this case, lol).

> > Does "assets" mean something different than how I'm using it? I can see that the word "client" would fit, but I never think of it as having a client side, since it's not a client-side webapp or anything like that.
> 
> Typically, "assets" means something like images or some non-code component to a project. At least that's been consistently true in my experience. While this isn't a "client side" app, that code is served specifically to a client, which is the meaning we want to convey to anyone else that may browse the source code. They'd see `src/client`, assume it was browser-side code, then come across its contents: HTML, CSS, and the JavaScript and the assumption would be validated. That's how I think of it anyway.

Makes perfect sense.

**Comment by Arlen22** [comment]
> > > I'm pretty sure tsc won't copy it
> > 
> > It will. I have it working on `master` I think. Maybe not, but I have it working somewhere :)
> 
> I'll dig around. That'd be news to me (which wouldn't surprise me in this case, lol).

A google search seems to indicate that tsc does not copy files over. I think using the webpack config will work fine.

**Comment by someguynamedmatt** [comment]
Yeah, it doesn't. Note the edit on that response. It transpiles the TypeScript _to_ the directory specified in the `tsconfig` file.

**Comment by Arlen22** [comment]
> Yeah, it doesn't. Note the edit on that response. It transpiles the TypeScript _to_ the directory specified in the `tsconfig` file.

Oh, right, Ok, makes sense.

**Comment by Arlen22** [comment]
@someguynamedmatt I rearranged the folders as you suggested. 

Everything compiles properly using `npm run build`.

**Comment by someguynamedmatt** [comment]
@Arlen22 on your master?

**Comment by Arlen22** [comment]
@someguynamedmatt  v2.2

**Comment by someguynamedmatt** [comment]
@Arlen22 Yeah, sorry. That's what I meant.

That branch doesn't build for me. It seems that the `build` command in `package.json`:

```
    "build": "rm -rf dist && tsc && webpack && npm run build:pkg"
```

is failing because `webpack` isn't listed as a `devDependency`. I'd bet this is working for you because you have some global install of webpack set up.

**Comment by someguynamedmatt** [comment]
Also, I noticed the file `build-package-json.js`. I'm not entirely sure what you're doing there, but it seems like a lib-version of the `package.json` is being created. I'd suggest [this Webpack plugin](https://www.npmjs.com/package/webpack-node-externals) is used instead. The only thing that should be blacklisted is the `devDependencies`.

**Comment by Arlen22** [comment]
> is failing because `webpack` isn't listed as a `devDependency`. I'd bet this is working for you because you have some global install of webpack set up.

Yes, good point. I use webpack with every project, so I don't install it in every project. You can just install it locally if you want.

**Comment by someguynamedmatt** [comment]
@Arlen22  Ok, but this is on the default branch of the project, right? You should push up the `devDependency` fix. I'd also like to suggest a few things:

1. Keep the `default` branch the same always (i.e. `master`) and only rely on release tags as you're doing. I think this will add a few benefits: 
    - The project will _always_ have a green branch
    - There won't be any confusion between developers (`master` is the default, well..."master branch" for 99% of projects)
    - There won't be the headache of diverging histories. `master` would just be tagged by releases, it can continually be merged into and when you're ready to release a new version of `TiddlyServer` you just `git tag -a v2.3 -m "new version released"` and that's it, no more needing to switch the repo's default branch

1. Get a build process going (see my `.travis.yml` file as an example)

So, if the project has a `master` branch which is always the canonical/working spearhead of the project you can protect it by gating it behind a build process (on TravisCI, say). This might be annoying if you're mostly working alone, but I think the slight inconvenience is worth it. Nothing gets onto master unless it (at a minimum) builds without failing. In this way the branch stays green. Ideally, the tests would be increased in the repository and they would also act as a gating mechanism to `master` merges. This also helps keep things sane when other people want to help out.

**Comment by Arlen22** [comment]
> Also, I noticed the file `build-package-json.js`. I'm not entirely sure what you're doing there, but it seems like a lib-version of the `package.json` is being created. I'd suggest [this Webpack plugin](https://www.npmjs.com/package/webpack-node-externals) is used instead. The only thing that should be blacklisted is the `devDependencies`.

The problem is we have three levels of dependencies. The `/src/server` (after the move) is actually designed to function as a library. The top server.ts is the only executable file. So dependencies is for that, and devDependencies is for the build process. At least that was the idea. What's your thoughts?

**Comment by someguynamedmatt** [comment]
I'm not sure I follow completely. Typically, with an npm lib you'd strip out the `devDependencies` since that's only relevant to a development team (but even that is sorta overkill), everything else in `dependencies` is required for the lib to function correctly so you want that bundled up. For instance, if you look at, say, `yargs`' `node_modules` in the `TiddlyServer` `node_modules` directory you'll see their `package.json` which lists everything. Conversely, React's `package.json` (I looked in a random project of my own) has `devDependencies` stripped out, probably just to keep internal tools a secret, I'd assume.

**Comment by Arlen22** [comment]
> I'm not sure I follow completely. Typically, with an npm lib you'd strip out the `devDependencies` since that's only relevant to a development team (but even that is sorta overkill), everything else in `dependencies` is required for the lib to function correctly so you want that bundled up.

`build-package-json.js` removes everything that isn't needed for the npm package. Since there is quite a bit of stuff that is needed in the GitHub repo, but much less for the NPM package, I thought this seemed like the best way to do it while still keeping the build and runtime dependencies separated.

**Comment by Arlen22** [comment]
> 1. Keep the `default` branch the same always (i.e. `master`) and only rely on release tags as you're doing. I think this will add a few benefits:

The main reason I did this was because I needed to keep the old branch around in case I need to do bug fixes on the old version. Mainly because each time I make a new branch its because I'm introducing incompatible changes. But I could branch the old one off master before merging the new one in.

**Comment by someguynamedmatt** [comment]
> The main reason I did this was because I needed to keep the old branch around in case I need to do bug fixes on the old version. Mainly because each time I make a new branch its because I'm introducing incompatible changes. But I could branch the old one off master before merging the new one in.

Ah, that makes sense. I hadn't considered that. I mean, I guess what you could do is have `master` always be the most updated code (i.e. stable, but unreleased) and anytime you release a new major (in the semver usage of that term) branch that off of master (like you're suggesting). So, roughly, it would be:

```
                       (v2.0)          (v3.0)
                      /                   /
(master branch)------/-------------------/-------------->
```

And then you could put fixes onto those branches as they're needed (minors and bug fixes) and tag them as needed (e.g. `v2.2`, etc). I think this is how bigger projects handle that stuff anyway. I'm just sorta thinking out loud.

**Comment by Arlen22** [comment]
Because I usually create a new branch with my changes and work on it for a while before merging it into master I'll branch the old one off then merge the new one in. That should take care of everything. 

I added webpack to devDependancies. I'll take a look at travis at some point, but it's naturally not high on my priority list since TypeScript catches most errors usually and opening a data folder catches the rest. If I ever add people to it I'll definitely want to do something like that. 

Feel free to reorganize `/src/server` whenever you have time. I have some opinions about how to organize things, but it's pretty easy to bounce off whatever you end up doing and tweak it if I don't like something. Do be aware that I am working on rewriting a lot of parts, so if you rather want to just give me some general ideas and industry standards, that's fine too. 

I do appreciate your help with all of this.

**Comment by someguynamedmatt** [comment]
@Arlen22 You’re going to need to add `webpack-cli` to that as well, fyi.

**Comment by Arlen22** [comment]
Done, thanks.

**Comment by Arlen22** [comment]
I moved the other deps into devDeps because I realized that you need almost all of the devDeps anyway in order for TypeScript to run, and I'm not going to support the library format in the near future. It's more of a design decision I might eventually turn into a library if TiddlyWiki ever gets big enough to create a demand for it, which is unlikely for data folders for multiple reasons.
